web/blog/matching-yara-rules-in-byte-pipelines/index.md

```diff
@@ -0,0 +1,200 @@
+0x4:3:$bar: bar
+0x7:3:$bar: bar
+```
+
+There are other ways to operationalize YARA rules, e.g., the
+[ClamAV](https://www.clamav.net/) scanner,
+[osquery](https://osquery.readthedocs.io/en/stable/deployment/yara/), or
+[Velociraptor](https://docs.velociraptor.app/vql_reference/plugin/yara/)—which
+we also [integrated as pipeline
+operator](/blog/integrating-velociraptor-into-tenzir-pipelines).
+
+## The `yara` operator
+
+Our new [`yara`][yara-operator] operator is a transformation that accepts bytes
+as input and produces events as output. Let's take the simple case of running
+the above example on string input:
+
+```bash
+echo 'foo barbar baz' |
+  tenzir 'load stdin | yara /tmp/test.yara'
+```
+
+The resulting `yara.match` events look as follows:
+
+```json
+{
+  "rule": {
+    "identifier": "test",
+    "namespace": "default",
+    "string": "foo",
+    "tags": [],
+    "meta": {
+      "string": "string meta data",
+      "integer": 42,
+      "boolean": true
+    },
+    "matches": [
+      {
+        "identifier": "$foo",
+        "data": "Zm9v",
+        "base": 0,
+        "offset": 0,
+        "match_length": 3
+      }
+    ]
+  }
+}
+{
+  "rule": {
+    "identifier": "test",
+    "namespace": "default",
+    "string": "bar",
+    "tags": [],
+    "meta": {
+      "string": "string meta data",
+      "integer": 42,
+      "boolean": true
+    },
+    "matches": [
+      {
+        "identifier": "$bar",
+        "data": "YmFy",
+        "base": 0,
+        "offset": 4,
+        "match_length": 3
+      },
+      {
+        "identifier": "$bar",
+        "data": "YmFy",
+        "base": 0,
+        "offset": 7,
+        "match_length": 3
+      }
+    ]
+  }
+}
+```
+
+Each event represents a rule match on a rule string. The output has a `rule`
+field that describes the metadata of the rule, and a `matches` array with a list
+of all matches for a given rule string. Each match includes the string
+`identifier`, e.g., `$foo`, an excerpt of the matching data in field `data` as
+blob,[^1] and details about the location of the match.
+
+[^1]: JSON doesn't distinguish binary blobs from strings. However, our type
+    system does, so we encode blob values as Base64-encoded strings for formats
+    that do not have a native blog representation.
+
+## Mixing and matching loaders
+
+The [`stdin`](/connectors/stdin) loader in the above example produces chunks of
+bytes. But you can use any connector of your choice that yields bytes. In
+particular, you can use the [`file`](/connectors/file) loader:
+
+```bash
+tenzir 'load file --mmap /tmp/test.txt | yara /tmp/test.yara'
+```
+
+Note the `--mmap` flag: we need it to deliver a single block of data to the
+`yara` operator. Without `--mmap`, the `file` loader will generate a stream of
+byte chunks and feed them incrementally to the `yara` operator, attempting to
+match `rule.yara` individually on *every chunk*. This will cause false negatives
+when rule matches span chunk boundaries.
```

Comment:
From the usability/safety point of view it would be better to make the `yara` operator failing if it receives a stream of byte chunks and adding an option like `--match-stream` to it if this is really needed.

Reply:
Good suggestion. I removed the need to provide `--mmap`. It's now only an optimization.

The default is now "buffer all memory blocks and initiate a scan at the end". I verified that the YARA scanner is capable of matching across block boundaries as follows:

```bash
# Spin up the scanner that matches on "$foo and $bar".
tenzir 'load zmq | yara rule.yara'

# Feed two blocks into the scanner:
echo foo | tenzir 'load stdin | to zmq'
echo bar | tenzir 'load stdin | to zmq'
```

A new `--blockwise` option allows to opt out of the accumulation and match on each byte chunk in isolation. This can run potentially forever without leaking memory.